### ISSUE TITLE
Unable to build protobuf/grpc

### ISSUE THREAD

**Issue by gouriano** [conan-io/conan#9919]
I am trying to build protobuf and grpc from sources and fail.

My OS: CentOS Linux release 7.9.2009
Compiler: GCC 7.3.0
I have the following conanfile.txt

`
    [requires]
    protobuf/3.17.1
    grpc/1.39.1
`

and run this command:
conan install . --build protobuf --build grpc --build missing

and it fails with messages like this:

build_subfolder/gens/src/proto/grpc/channelz/channelz.pb.cc:563:76: error: temporary of non-literal type 'grpc::channelz::v1::GetSocketRequest' in a constant expression

I believe it is a code generated by protoc.
Is it a problem with protobuf or with my setup?

Thanks

**Comment by gouriano** [comment]
Maybe it requires a newer compiler?
The same setup works just fine on Windows, with Visual Studio 2019

**Comment by memsharded** [comment]
If the recipe requires a more modern compiler, then the typical recipe should raise a ``ConanInvalidConfiguration`` error to indicate such

**Comment by uilianries** [comment]
Yes, it requires a newer version, but still your error message is not accurate and doesn't help you. The recipe should raise `ConanInvalidConfiguration` with something more helpful.

**Comment by SSE4** [comment]
according to their docs: https://github.com/grpc/grpc/tree/master/src/cpp
```
Linux - Debian, Ubuntu, CentOS | x86, x64 | clang 4+, GCC 4.9+ | Officially Supported
```

they claim GCC 4.9+ on CentOS is supported, so GCC 7.3.0 should be also supported.

**Comment by vakatov** [comment]
@gouriano - FWIW, I just tried building it from scratch, and it seems to have succeeded:

```
$ conan install . --build protobuf --build grpc --build missing
.....................
-- Installing: /home/vakatov/.conan/data/grpc/1.39.1/_/_/package/43b71c9e3ad1cd89fe61dc726ebe73acbd18d0d4/lib/pkgconfig/grpc++_unsecure.pc
grpc/1.39.1 package(): Packaged 12 '.a' files
grpc/1.39.1 package(): Packaged 1 '.cmake' file: grpc_cpp_plugin.cmake
grpc/1.39.1 package(): Packaged 252 '.h' files
grpc/1.39.1 package(): Packaged 8 files
grpc/1.39.1: Package '43b71c9e3ad1cd89fe61dc726ebe73acbd18d0d4' created
grpc/1.39.1: Created package revision 4aeccf5ecd200f80c4bafc44c55b45ff
grpc/1.39.1: Appending PATH environment variable: /home/vakatov/.conan/data/grpc/1.39.1/_/_/package/43b71c9e3ad1cd89fe61dc726ebe73acbd18d0d4/bin
conanfile.txt: Generator txt created conanbuildinfo.txt
conanfile.txt: Aggregating env generators
conanfile.txt: Generated conaninfo.txt
conanfile.txt: Generated graphinfo

$ echo $?
0
```

NOTE: I had to call this though (as Conan recommended):
`$  conan profile update settings.compiler.libcxx=libstdc++11 default`

So I reckon we should re-investigate this problem on-site, stall this CONAN ticket meanwhile.

**Comment by gouriano** [comment]
Found it. I had

    [settings]
    compiler.cppstd=17

With compiler.cppstd=14 it builds fine

**Comment by gouriano** [comment]
Still, I believe GRPC/PROTOBUF should be fixed. Both recipe and the packages.
Recipe should take care of the language standard
Packages... Well, PROTOC generates (at least in some scenarios) code which cannot be compiled with C++17. And our own package cannot be compiled without C++17.
So, if we need PROTOC code generation (and we do need it) we could end up in situation when our package cannot be compiled.

**Comment by uilianries** [comment]
@gouriano Thank you for your explanation! Packages which are cppstd dependent usually we validate it under `validate()` method, probably this one doesn't have such validation.
